### ISSUE TITLE
Slovenian eOI on Mac only exposes the PIN-less applet

### ISSUE THREAD

**Issue by craftbyte** [OpenSC/OpenSC#2986]
### Problem Description
The Slovenian eID implementation on Mac seems to only expose the PIN-less applet to the OS. Running `pkcs11-tool --list-slots` works though. Seems like there is something weird between the TokenD service and OpenSC.
Pinging @llogar, since he was the main implementor of this and may have some details.

### Proposed Resolution
Expose both applets to the OS. Maybe we need to present all the certificates as one token?

### Steps to reproduce
Insert eOI into Mac. Run `security list-smartcards` and `security export-smartcard`. The output will only show the PIN-less applet details.

### Logs
https://gist.github.com/craftbyte/9be5e9c0b9d319880eac7fb6d13ef289

**Comment by llogar** [comment]
Currently only one applet at a time can be active. I think this is an **OpenSC** limitation. To select an active applet, enable/disable the relevant application block in `opensc.conf` (either `E828BD080F014E585031` or `E828BD080F014E585030`)`. However I think the pinless applet should, by default, be **disabled**, so I am not sure why you are only seeing the pinless applet/slot

**Comment by frankmorgner** [comment]
I think this is a limitation of OpenSCToken, which expects all keys and certificates to be in the *generic* card application:
https://github.com/frankmorgner/OpenSCToken/blob/f860cabca2d99bd600eb2affb2a8ef0a2a9b4bc0/OpenSCToken/Token.m#L106-L108

Basically, that needs to be extended to the other applications on the card as well, which is similar to what is done in the PKCS#11 library:
https://github.com/OpenSC/OpenSC/blob/c3545018d059b4debde33b9f34de719dd41e5531/src/pkcs11/slot.c#L347-L361

Interestingly, minidriver.c also [only binds the *generic* application](https://github.com/OpenSC/OpenSC/blob/c3545018d059b4debde33b9f34de719dd41e5531/src/minidriver/minidriver.c#L7142-L7153). Could you please check if this problem also occurs on Windows using `certutil.exe -scinfo`?

**Comment by llogar** [comment]
Yes, it's the same on Windows. As I've said before, the solution (at least on Windows, I have no experience with Macs) is to enable just the applet you need...

**Comment by frankmorgner** [comment]
I added multi-app support into OpenSCToken. Unfortunately, I can only do some basic tests. Would you mind testing the macOS package from here https://github.com/OpenSC/OpenSC/actions/runs/7653681110 ?

**Comment by frankmorgner** [comment]
Would someone please test the macOS artifacts linked above? This adds support for all applications on the card, without the need for the mentioned workaround. thank you

**Comment by msetina** [comment]
Should this work on Linux?

**Comment by craftbyte** [comment]
@frankmorgner Sorry for the late reply, I have been hit with life as a truck. I tested this and am having issues with basic OpenSC usage. MacOS doesnt detect any smartcards and pkcs11-tool takes a suspiciously long time to list slots. What kind of logs would be useful to you?

**Comment by llogar** [comment]
> Should this work on Linux?

It should... The pinless applet is disabled by default though and when/if enabled it erroneously asks for PIN, although it is not needed. See https://github.com/OpenSC/OpenSC/pull/2646#discussion_r1025129143

**Comment by msetina** [comment]
Thank you @llogar. I was wondering about this modification @frankmorgner linked, because I was following this support and as I can see there are two features missing as "by desing as of now" in openSC:
1. PINless is not there but with circumvention
2. Support is there for just one app per card and the forementioned fix should add this. So all 3 apps on the card need to be enabled on config. One at a time.

**Comment by llogar** [comment]
There are only 2 applets of interest (well, the 3rd one is eMRTD applet, but I think it's irrelevant in this context). If both are enabled there are 3 virtual slots (1 slot for pinless applet and 2 slots for the signature applet (1 for NormPIN and 1 for SigPIN). I prefer to have only one applet enabled at a time, as If I remember correctly, firefox (or perhaps thunderbird) kept nagging me for PIN entry for the unneeded one. But ymmv.

**Comment by frankmorgner** [comment]
> I added multi-app support into OpenSCToken. Unfortunately, I can only do some basic tests. Would you mind testing the macOS package from here https://github.com/OpenSC/OpenSC/actions/runs/7653681110 ?

This modification should allow using all certificates from all applets in OpenSCToken at the same time without modifying the active applet in opensc.conf. To test this, download the build artifact (https://github.com/OpenSC/OpenSC/actions/runs/7653681110/artifacts/1194532843), install the dmg. `sc_auth identities` should now show all certificates.

**Comment by msetina** [comment]
@craftbyte maybe a fix to your problem is just adding:
```
framework pkcs15 {
                # Slovenian eID - low level (pinless, "Prijava brez PIN-a")
                application E828BD080F014E585031 {
                model = "ChipDocLite";
                disable = true;
                user_pin = "Norm PIN";
                }

                # Slovenian eID - high level (QES, "Podpis in prijava")
                application E828BD080F014E585030 {
                model = "ChipDocLite";
                # disable = true;
                user_pin = "Norm PIN";
                sign_pin = "Sig PIN";
                }
        }

```
to app default section of opensc.conf. This is from example opensc.conf and it works for me.

**Comment by frankmorgner** [comment]
@llogar , the settings you are referring to are present in `opensc.conf.example`, but they are not distributed as `opensc.conf`, i.e. they are *not the default* settings. Ideally, we'd like to keep it that way so that OpenSC runs with sane settings even when a configuration file is missing. 

There are already tweaks to the PKCS#15 profile in `pkcs15-eoi.c` and if possible, this is where you can modify the profile to work out of the box. For example, you could easily modify `sc_pkcs15emu_eoi_init_ex` to return an error if binding `E828BD080F014E585031` is requested (also with a driver specific option to maybe enable this again). Manually renaming *Norm PIN" to "UserPIN" and *Sig PIN* to "SignPIN" could also be an option, but I don't know if it is required to be identified correctly.

**Comment by frankmorgner** [comment]
What's the status of this topic, is there anything to do?
